syntax = "proto3";

package brain.action.v1;

import "brain/shared/v1/envelope.proto";

service AttentionRouterService {
  rpc RouteNotification(RouteNotificationRequest) returns (RouteNotificationResponse);
  rpc RouteApprovalNotification(RouteApprovalNotificationRequest) returns (RouteNotificationResponse);
  rpc FlushBatch(FlushBatchRequest) returns (RouteNotificationResponse);
  rpc CorrelateApprovalResponse(CorrelateApprovalResponseRequest) returns (CorrelateApprovalResponseResponse);
  rpc Health(AttentionRouterHealthRequest) returns (AttentionRouterHealthResponse);
}

message RoutedNotification {
  string actor = 1;
  string channel = 2;
  string recipient = 3;
  string sender = 4;
  string message = 5;
  string title = 6;
  string dedupe_key = 7;
  string batch_key = 8;
}

message RouteNotificationResult {
  string decision = 1;
  bool delivered = 2;
  string detail = 3;
  string suppressed_reason = 4;
  int32 batched_count = 5;
  RoutedNotification notification = 6;
}

message RouteNotificationPayload {
  string actor = 1;
  string channel = 2;
  string title = 3;
  string message = 4;
  string recipient_e164 = 5;
  string sender_e164 = 6;
  string dedupe_key = 7;
  string batch_key = 8;
  bool force = 9;
}

message RouteNotificationRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  RouteNotificationPayload payload = 2;
}

message PolicyApprovalNotificationPayload {
  string proposal_token = 1;
  string capability_id = 2;
  string capability_version = 3;
  string summary = 4;
  string actor = 5;
  string channel = 6;
  string trace_id = 7;
  string invocation_id = 8;
  string expires_at = 9;
}

message RouteApprovalNotificationRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  PolicyApprovalNotificationPayload payload = 2;
}

message FlushBatchPayload {
  string batch_key = 1;
  string actor = 2;
  string channel = 3;
  string recipient_e164 = 4;
  string sender_e164 = 5;
  string title = 6;
}

message FlushBatchRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  FlushBatchPayload payload = 2;
}

message ApprovalCorrelationPayload {
  string actor = 1;
  string channel = 2;
  string message_text = 3;
  string approval_token = 4;
  string reply_to_proposal_token = 5;
  string reaction_to_proposal_token = 6;
}

message CorrelateApprovalResponseRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  ApprovalCorrelationPayload payload = 2;
}

message CorrelateApprovalResponseResponse {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  ApprovalCorrelationPayload payload = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message RouteNotificationResponse {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  RouteNotificationResult payload = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message AttentionRouterHealthPayload {}

message AttentionRouterHealthRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  AttentionRouterHealthPayload payload = 2;
}

message AttentionRouterHealthStatus {
  bool service_ready = 1;
  bool adapter_ready = 2;
  string detail = 3;
}

message AttentionRouterHealthResponse {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  AttentionRouterHealthStatus payload = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}
